File: Leetcode 1930. Unique Length-3 Palindromic Subsequences/lc 1930.java
class Solution {
    public int countPalindromicSubsequence(String s) {
        int ans=0;
        HashMap<Character,int[]> hm=new HashMap<>();
        for(int i=0;i<s.length();i++){
            char c=s.charAt(i);
            if(hm.containsKey(c)){
                int[] pair=hm.get(c);
                pair[1]=i;
            }else{
                int[] pair={i,-1};
                hm.put(c,pair);
            }
        }

        for(Map.Entry<Character,int[]> m:hm.entrySet()){
            char c=m.getKey();
            int[] arr=m.getValue();
            if(arr[1]>0){
                Set<Character> set=new HashSet<>();
                for(int i=arr[0]+1;i<arr[1];i++){
                    set.add(s.charAt(i));
                }
                ans+=set.size();
            }
        }
        return ans;
    }
}

File: Leetcode 1930. Unique Length-3 Palindromic Subsequences/lc 1930.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        int ans=0;
        unordered_map<char,pair<int,int>> mp;

        for(char c='a';c<='z';c++){
            mp[c]={-1,-1};
        }

        for(int i=0;i<s.length();i++){
            if(mp[s[i]].first==-1){
                mp[s[i]].first=i;
            }else{
                mp[s[i]].second=i;
            }
        }

        for(auto m:mp){
            int sta=m.second.first;
            int e=m.second.second;

            if(e!=-1){
                unordered_set<char> st;
                for(int i=sta+1;i<e;i++){
                    st.insert(s[i]);
                }
                ans+=st.size();
            }
            
        }
        
        return ans;
    }
};